### PR TITLE
pkg/destroy/azure: Delete app registrations during cluster destroy

### PR DESCRIPTION
**Pull request by joelddiaz**
With cloud-cred-operator tagging Service Principals, we can now located the Service Principals by tag, and find their parent Application Registration and delete them as part of cluster destroy.

This should keep the App Registrations created by cloud-cred-operator from leaking out after cluster create/destroy.

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
dont' log errors with ERROR level, the destroy for azure returns them and prints them in DEBUG.

File: pkg/destroy/azure/azure.go [hunk 1]
```diff
@@ -24,27 +25,37 @@ import (
...
+	zonesClient             dns.ZonesClient
+	recordsClient           dns.RecordSetsClient
+	serviceprincipalsClient graphrbac.ServicePrincipalsClient
+	applicationsClient      graphrbac.ApplicationsClient
 }
 
 func (o *ClusterUninstaller) configureClients() {
 	o.resourceGroupsClient = resources.NewGroupsGroupClient(o.SubscriptionID)
-	o.resourceGroupsClient.Authorizer = o.Authorizer
+	o.resourceGroupsClient.Authorizer = o.ResourceManagerAuthorizer
```
Comment (abhinavdahiya) [inline pkg/destroy/azure/azure.go]: hmm.. this should be `Authorizer`

File: pkg/destroy/azure/azure.go [hunk 2]
```diff
@@ -217,3 +235,68 @@ func deleteResourceGroup(ctx context.Context, client resources.GroupsGroupClient
...
+	tag := fmt.Sprintf("kubernetes.io_cluster.%s=owned", infraID)
+	servicePrincipals, err := getServicePrincipalsByTag(ctx, spClient, tag, infraID)
+	if err != nil {
+		return errors.Wrap(err, "failed to gather list of Service Principals by tag")
+	}
+
+	for _, sp := range servicePrincipals {
+		logger = logger.WithField("appID", *sp.AppID)
+		appFilter := fmt.Sprintf("appId eq '%s'", *sp.AppID)
+		appResults, err := appClient.List(ctx, appFilter)
```
Comment (abhinavdahiya) [inline pkg/destroy/azure/azure.go]: why not use https://godoc.org/github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac#ApplicationsClient.Get vs listing for appID can there be more than one with same ID???

File: pkg/destroy/azure/azure.go [hunk 3]
```diff
@@ -217,3 +235,68 @@ func deleteResourceGroup(ctx context.Context, client resources.GroupsGroupClient
...
+		logger.Info("deleted")
+	}
+
+	return utilerrors.NewAggregate(errorList)
+}
+
+func getServicePrincipalsByTag(ctx context.Context, spClient graphrbac.ServicePrincipalsClient, matchTag, infraID string) ([]graphrbac.ServicePrincipal, error) {
+	matchedSPs := []graphrbac.ServicePrincipal{}
+
+	infraFilter := fmt.Sprintf("startswith(displayName,'%s')", infraID)
```
Comment (abhinavdahiya) [inline pkg/destroy/azure/azure.go]: use somethings like https://github.com/openshift/installer/blob/045402165b7aa5cb95ab050d4ca0827016e05dd1/pkg/asset/installconfig/azure/dns.go#L121-L132

File: pkg/destroy/azure/azure.go [hunk 4]
```diff
@@ -217,3 +235,68 @@ func deleteResourceGroup(ctx context.Context, client resources.GroupsGroupClient
...
+	spResults, err := spClient.List(ctx, infraFilter)
+	if err != nil {
+		return matchedSPs, err
+	}
+
+	for {
```
Comment (abhinavdahiya) [inline pkg/destroy/azure/azure.go]: use somethings like https://github.com/openshift/installer/blob/045402165b7aa5cb95ab050d4ca0827016e05dd1/pkg/asset/installconfig/azure/dns.go#L121-L132

File: pkg/destroy/azure/azure.go (hunk 2 above)
Comment (joelddiaz) [reply pkg/destroy/azure/azure.go]: the Get requires querying the objectID not the appID
```
[jdiaz@minigoomba ~]$ az ad app list --display-name jdiaz-app | jq -r ".[].appId , .[].objectId"
9ba357ad-0bf5-4a0c-8c3f-8e6eee8e6b36
10dbbb0c-29ae-4705-8bf9-95e99f0e8a3b
```
And all you get from the service principal is the appID of the app registration.

File: pkg/destroy/azure/azure.go (hunk 4 above)
Comment (joelddiaz) [reply pkg/destroy/azure/azure.go]: :+1: i much prefer this. will change it

File: pkg/destroy/azure/azure.go (hunk 1 above)
Comment (joelddiaz) [reply pkg/destroy/azure/azure.go]: fixed

**Comment by abhinavdahiya** [comment]
/test e2e-azure

LGTM

con you squash the commits 80ca6142c8c94b32f69ab07b48c7f3c706e524cb and 96d22b6c05bdedb79dafd48338d51a8f22867674

**Comment by joelddiaz** [comment]
squashed

**Comment by ingvagabund** [comment]
> I think it was done out of concern for the 93 character limit. @ingvagabund maybe you can give more context?

Azure does not allow application )SP) names to be longer than 93 characters. It errors otherwise.

**Comment by joelddiaz** [comment]
looks like it's deleting the app registrations now
```
time="2019-08-27T18:08:30Z" level=debug msg="deleting application registrations"
time="2019-08-27T18:08:31Z" level=info msg=deleted appID=f18a9b1f-72ff-43db-8a73-abf73bc37ad9
time="2019-08-27T18:08:37Z" level=info msg=deleted appID=47c260de-9275-4db2-b889-c34334c33ee9
time="2019-08-27T18:08:43Z" level=info msg=deleted appID=2e39df42-0018-472c-b1b0-20c524eb2d3c
```
just need to get through CI now...

**Pull request merged**